File: solutions/025.Reverse_Nodes_in_k-Group/AC_recursive_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_recursive_n.java
*  Create Date: 2015-03-05 14:59:35
*  Descripton:   
*/

import java.util.*;

public class Solution {

    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode cur = head;
        int cnt = 0;
        // get next group
        while (cur != null && cnt != k) {
            cur = cur.next;
            cnt++;
        }
        if (cnt == k) {
            cur = reverseKGroup(cur, k);
            // reverse
            while (0 <= --cnt) {
                ListNode tmp = head.next;
                head.next = cur;
                cur = head;
                head = tmp;
            }
            head = cur;
        }
        return head;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 4};
        System.out.println("no case");
    }
}

File: solutions/025.Reverse_Nodes_in_k-Group/AC_recursive_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_recursive_n.cpp
*  Create Date: 2015-01-15 20:45:27
*  Descripton:  Count k nodes and recursive
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        if (!head || !(head->next) || k < 2)
            return head;
        
        // count k nodes
        ListNode *nextgp = head;
        for (int i = 0; i < k; i++)
            if (nextgp)
                nextgp = nextgp->next;
            else
                return head;

        // reverse
        ListNode *prev = NULL, *cur = head, *next = NULL;
        while (cur != nextgp) {
            next = cur->next;
            if (prev)
                cur->next = prev;
            else
                cur->next = reverseKGroup(nextgp, k);
            prev = cur;
            cur = next;
        }
        return prev;
    }
};

int main() {

    return 0;
}

